src/history/hs_rec.c

```diff
@@ -741,32 +719,29 @@ __hs_fixup_out_of_order_from_pos(WT_SESSION_IMPL *session, WT_CURSOR *hs_cursor,
 #ifndef HAVE_DIAGNOSTIC
     WT_UNUSED(key);
 #endif
-    /*
-     * Position ourselves at the beginning of the key range that we may have to fixup. Prior to
-     * getting here, we've positioned our cursor at the end of a key/timestamp range and then done a
-     * "next". Normally that would leave us pointing at higher timestamps for the same key (if any)
-     * but in the case where our insertion timestamp is the lowest for that key, our cursor may be
-     * pointing at the previous key and can potentially race with additional key insertions. We need
-     * to keep doing "next" until we've got a key greater than the one we attempted to position
-     * ourselves with.
-     */
+
+    /* If we will delete all the updates of the key from the history store, we should not reinsert
+     * any update. */
+    WT_ASSERT(session, ts > WT_TS_NONE || !reinsert);
+
     for (; ret == 0; ret = hs_cursor->next(hs_cursor)) {
         /* We shouldn't have crossed the btree and user key search space. */
         WT_ERR(hs_cursor->get_key(hs_cursor, &hs_btree_id, &hs_key, &hs_ts, &hs_counter));
-        WT_ASSERT(session, hs_btree_id == btree->id);
+        WT_ASSERT(session, hs_btree_id == btree_id);
 #ifdef HAVE_DIAGNOSTIC
         WT_ERR(__wt_compare(session, NULL, &hs_key, key, &cmp));
         WT_ASSERT(session, cmp == 0);
 #endif
-        if (hs_ts > ts)
+        /* We find a key that is larger or equal to the specified timestamp*/
+        if (hs_ts >= ts)
             break;
     }
     if (ret == WT_NOTFOUND)
         return (0);
     WT_ERR(ret);
 
     /*
-     * The goal of this fixup function is to move out-of-order content to maintain ordering in the
+     * The goal of this function is to move out-of-order content to maintain ordering in the
```

Comment:
Do we want to add an example of how we deal with mixed mode timestamps here?

Reply:
Added